src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int Count<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+            
+            IEnumerator<T> enumerator = source.GetEnumerator();
+            return CountAndDispose(enumerator, enumerator);
+        }
+
+        /// <summary>
+        /// Gets the number of items in the enumerator, and then disposes it.
+        /// </summary>
+        /// <param name="enumerator">The enumerator to count.</param>
+        /// <param name="disposable">The enumerator typed as an <see cref="IDisposable"/>.</param>
+        /// <returns>The count of the enumerator.</returns>
+        /// <remarks>This method should be kept non-generic to avoid generated code bloat.</remarks>
```

Comment:
> We only go down this path if the enumerable is lazy

This is extremely common.  If you have a concrete ```T[]```, you're much better off calling Length.  If you have a ```List<T>``` or an ```ICollection<T>``` or an ```IList<T>```, etc., you're better off just using its Count property.  It's only when you have one of those that's typed as ```IEnumerable<T>``` that you'd use Count(), whereas Count() is the only option for enumerables produced by LINQ, by iterators, etc.

> Separating this part out into a new method saves the JIT from generating a sizeable chunk of code for each generic instantiation of Count

I understand.  But at the same time, we're not going to go through every generic method in the platform and pull out groups of lines here and there that aren't dependent on the generic parameters and separate them into their own methods.  That would be overkill.  What makes these few lines special?